feat: add event scheduler by json

dudantas opened the pull request:
This pull request introduces a new JSON-based event scheduler system for in-game events, adds initial event data and documentation, and integrates event-driven gameplay effects into the server logic. It also updates the build configuration to require the `nlohmann_json` library. The most important changes are grouped below:

**Event Scheduler System and Data:**

* Added a new `events.json` file with multiple scheduled in-game events, each specifying properties such as start/end dates, rates (exp, loot, spawn), and special effects (e.g., double bestiary, fast exercise, forge chance).
* Provided a comprehensive `README.md` in the event scheduler directory, documenting the structure and usage of the `events.json` configuration for custom events.
* Added a sample script (`example.lua`) for event scheduling integration.

**Server Integration and Logic:**

* Updated server module loading to support both XML and JSON event schedules, ensuring JSON events are loaded after scripts for scripting interface compatibility. [[1]](diffhunk://#diff-2f9a824bc89a72752ae30282d88f40c8307265f1b37189f7b27cced21846c5afL389) [[2]](diffhunk://#diff-2f9a824bc89a72752ae30282d88f40c8307265f1b37189f7b27cced21846c5afR415-R418)
* Integrated event-driven effects into gameplay:
  - Players receive logging about active events upon login.
  - Double bestiary and bosstiary event effects are applied to kill counters, with logging for debugging. [[1]](diffhunk://#diff-446c95b161e562fe80eee30b6f967d7bfaa0a69d6c2ed46e2c5ba7a4ba8d9f9aR6146-R6153) [[2]](diffhunk://#diff-446c95b161e562fe80eee30b6f967d7bfaa0a69d6c2ed46e2c5ba7a4ba8d9f9aR6165-R6172)
  - Forge success rate is adjusted based on event configuration.
  - Exercise training weapon speed is doubled if the fast exercise event is active.

**Build System Updates:**

* Added `nlohmann_json` as a required dependency in the CMake configuration and linked it to the project libraries. [[1]](diffhunk://#diff-4166551a5ac28cf3eaeff778e899b408f3ecf6d667eb0afabe3c7f9069ab0edeR96-R100) [[2]](diffhunk://#diff-067aa149c1df301dfba501917d0a66fc57ad7f66269d1e41676f516edf16a289R110)

**Codebase Maintenance:**

* Included the event scheduler header in the player implementation for direct event querying.

These changes lay the foundation for flexible, data-driven event scheduling and dynamic gameplay modifications based on active events.

New case from login.php (myaac):
```php
case 'eventschedule':
        $eventlist = [];

        $json_file_path = config('server_path') . 'data/json/eventscheduler/events.json';
        $xml_file_path = config('server_path') . 'data/XML/events.xml';

        if (file_exists($json_file_path)) {
            $jsonContent = file_get_contents($json_file_path);
            $events = json_decode($jsonContent, true);

            if (json_last_error() !== JSON_ERROR_NONE) {
                error_log("Error parsing JSON: " . json_last_error_msg());
                die(json_encode([]));
            }

            foreach ($events['events'] as $event) {
                if ($event) {
                    $startdate = strtotime($event['startdate'] ?? ''); // Convert start date to timestamp
                    $enddate = strtotime($event['enddate'] ?? '');     // Convert end date to timestamp

                    $eventData = [
                        'colorlight'      => $event['colors']['colorlight'] ?? '',
                        'colordark'       => $event['colors']['colordark'] ?? '',
                        'description'     => $event['description'] ?? '',
                        'displaypriority' => intval($event['details']['displaypriority'] ?? 0),
                        'enddate'         => $enddate,
                        'isseasonal'      => getBoolean(intval($event['details']['isseasonal'] ?? 0)),
                        'name'            => $event['name'] ?? '',
                        'startdate'       => $startdate,
                        'specialevent'    => getBoolean(intval($event['details']['specialevent'] ?? 0))
                    ];
                    $eventlist[] = $eventData;
                }
            }

        } elseif (file_exists($xml_file_path)) {
            $xml = new DOMDocument;
            $xml->load($xml_file_path);
            $tableevent = $xml->getElementsByTagName('event');

            foreach ($tableevent as $event) {
                if ($event) {
                    $eventData = [
                        'colorlight'      => parseEvent($event->getElementsByTagName('colors'), false, 'colorlight'),
                        'colordark'       => parseEvent($event->getElementsByTagName('colors'), false, 'colordark'),
                        'description'     => parseEvent($event->getElementsByTagName('description'), false, 'description'),
                        'displaypriority' => intval(parseEvent($event->getElementsByTagName('details'), false, 'displaypriority')),
                        'enddate'         => strtotime(parseEvent($event, true, false)), // Convert end date to timestamp
                        'isseasonal'      => getBoolean(intval(parseEvent($event->getElementsByTagName('details'), false, 'isseasonal'))),
                        'name'            => $event->getAttribute('name'),
                        'startdate'       => strtotime(parseEvent($event, true, true)), // Convert start date to timestamp
                        'specialevent'    => getBoolean(intval(parseEvent($event->getElementsByTagName('details'), false, 'specialevent')))
                    ];
                    $eventlist[] = $eventData;
                }
            }
        } else {
            die(json_encode([]));
        }

        error_log("Event list prepared: " . print_r($eventlist, true));
        error_log("Last update timestamp: " . time());

        $response = [
            'eventlist' => $eventlist,
            'lastupdatetimestamp' => time()
        ];
        header('Content-Type: application/json');
        echo json_encode($response);
        break;
```

### Feature Highlights:
- New JSON-based event scheduler configuration.
- Automatic loading of event details such as name, start and end dates, and loot settings from a JSON file.
- Backward-compatible approach that ensures current functionality remains intact.

## Behaviour

### **Actual**

Currently, the Event Scheduler uses an XML file to manage events, which makes it more difficult to add or modify configurations due to the verbosity and inflexibility of XML.

### **Expected**

The Event Scheduler should now use a JSON file (`events.json`) to manage events. JSON provides a more compact, readable, and maintainable structure for event configuration. Users should be able to add, edit, and manage events more easily with this new format.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested

Tests were conducted to ensure that the new JSON format loads correctly and events are scheduled as expected. The tests included validating successful parsing of the JSON file, checking the correct initialization of event properties, and ensuring existing functionality remains stable.

- [x] Loaded various JSON configurations with multiple events to verify accurate parsing.
- [x] Confirmed that events are correctly registered, including start and end dates, and associated loot tables.
- [x] Tested with malformed JSON to ensure appropriate error handling.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I checked the PR checks reports
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my feature works effectively

dudantas committed:
feat: add event scheduler by json

dudantas committed:
feat: send events bonus

github-actions[bot] committed:
Lua code format - (Stylua)

dudantas committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json

dudantas committed:
improve: eventscheduler

murilo09 committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json

Codemaufeito commented:
can it be tested?

murilo09 commented:
> can it be tested?

yes

majestyotbr committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json

Codemaufeito reviewed (commented):
![image](https://github.com/user-attachments/assets/ff29ff9f-a466-4f0d-ac77-a37a7392dbcf)

porem quando o jogador efetuar o login não informa no server log

![image](https://github.com/user-attachments/assets/85bb9207-59af-42f3-a6c9-0ae28c58de90)

parece que não está ativo

no event.json está configurado como no exemplo

		{
			"name": "50% Loot Bonus",
			"startdate": "01/01/2025",
			"enddate": "01/17/2025",
			"ingame": {
				"lootrate": 150
			},
			"description": "Increases loot by 50%.",
			"colors": {
				"colordark": "#2b1e10",
				"colorlight": "#4a2e18"
			},
			"details": {
				"displaypriority": 6,
				"isseasonal": 0,
				"specialevent": 1
			}
		},

majestyotbr committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json

dudantas committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json

dudantas committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json

github-actions[bot] committed:
Code format - (CMake-format)

gabrielew commented:
@dudantas can add support to hours? 
```json
"startdate": "01/01/2025",
"enddate": "01/17/2025",
"hour": "21:00"
```

Because we have SS (Server Save) at very different times, to properly handle an “end date” for example, one that should end on the 30th, we actually need to set the event to end one day earlier. That’s because when the SS occurs at 21:00, the system still considers the current date to be the 30th.

dudantas commented:
> @dudantas can add support to hours?
> 
> ```json
> "startdate": "01/01/2025",
> "enddate": "01/17/2025",
> "hour": "21:00"
> ```
> 
> Because we have SS (Server Save) at very different times, to properly handle an “end date” for example, one that should end on the 30th, we actually need to set the event to end one day earlier. That’s because when the SS occurs at 21:00, the system still considers the current date to be the 30th.

Yes, I can see that

dudantas committed:
Add EventsScheduler JSON loading and active event tests

Refactored EventsScheduler to improve JSON event loading, handle duplicate event rates, and provide a reset method. Added getActiveEvents() to retrieve currently active events. Introduced comprehensive unit tests for event scheduling, duplicate handling, and state reset, along with necessary test fixtures and configuration files.

dudantas committed:
Remove events.json and update event handling logic

Deleted the obsolete data/json/events.json file and updated event references in eventscheduler/events.json, changing 'doubleexercise' to 'fastexercise'. Improved forge success rate calculation in game.cpp to clamp values between 0 and 100. Fixed protocolgame.cpp to default schedulePercentage to 100 if zero, ensuring correct cooldown timer calculations.

dudantas commented:
@codex review

dudantas committed:
Add hour-level event scheduling support

Refactored EventsScheduler to support event start and end times with hour and second precision, instead of only by day. Updated JSON and XML parsing to handle 'starthour', 'endhour', and 'hour' fields, and added validation for time ranges. Adjusted EventScheduler struct to use std::time_t for start and end times. Added unit test to verify correct handling of event hours from JSON.

dudantas committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json

github-actions[bot] committed:
Code format - (Clang-format)

dudantas commented:
> @dudantas can add support to hours?
> 
> ```json
> "startdate": "01/01/2025",
> "enddate": "01/17/2025",
> "hour": "21:00"
> ```
> 
> Because we have SS (Server Save) at very different times, to properly handle an “end date” for example, one that should end on the 30th, we actually need to set the event to end one day earlier. That’s because when the SS occurs at 21:00, the system still considers the current date to be the 30th.

Test this please: https://github.com/opentibiabr/canary/pull/3085/commits/9ee1c61907b424baf1358e90ef2ee0db3f1174f4

dudantas committed:
Fix event script path and correct event rate storage

Updated the script file path in EventsScheduler to use the correct 'json/eventscheduler/scripts' directory. Changed loadedScripts to use std::string instead of std::string_view for compatibility. Fixed the storage of 'forge-chance' and 'boss-cooldown' to use the actual event rates instead of subtracting 100. Updated related unit tests to match the new expected values.

dudantas committed:
Add integration tests for EventsScheduler

Introduces integration tests for the EventsScheduler, including a new test fixture for Lua environments and dependency injection improvements. Adds tests for event loading, stacking, time window filtering, and modifier clearing. Updates CMake and test setup to support new tests and fixtures.

github-actions[bot] committed:
Code format - (Clang-format)

github-actions[bot] committed:
Code format - (CMake-format)

dudantas committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json

dudantas committed:
Change logger level for double bestiary and bosstiary

dudantas committed:
Change log level from info to debug for bosstiary

dudantas committed:
Refactor event scheduler tests to use shared helpers

Introduced 'events_scheduler_test_helpers.hpp' with reusable test helper functions for event scheduler tests. Updated integration and unit tests to use these helpers, improving readability and reducing code duplication.

dudantas committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json

dudantas committed:
Refactor events scheduler test setup into shared fixture

Extracted common test setup logic for events scheduler tests into a shared EventsSchedulerTestBase fixture in tests/shared/game/events_scheduler_test_fixture.hpp. Updated integration and unit tests to inherit from this base class, reducing code duplication and improving maintainability. Also added [[nodiscard]] attributes to several Player class methods for better static analysis.

dudantas committed:
Merge branch 'main' into dudantas/improve-eventscheduler-by-json
